### PR TITLE
KAFKA-3234; Clarify minISR in documentation and auto-generate topic configuration docs

### PR DESCRIPTION
**Pull request by jjkoshy**
Fix design doc comments to clarify minISR and unclean leader election, switch topic configuration doc to be auto-generated.

### CODE REVIEW

File: docs/configuration.html [hunk 1]
```diff
@@ -52,100 +52,7 @@
...
-      <td>min.cleanable.dirty.ratio</td>
-      <td>0.5</td>
-      <td>log.cleaner.min.cleanable.ratio</td>
-      <td>This configuration controls how frequently the log compactor will attempt to clean the log (assuming <a href="#compaction">log compaction</a> is enabled). By default we will avoid cleaning a log where more than 50% of the log has been compacted. This ratio bounds the maximum space wasted in the log by duplicates (at 50% at most 50% of the log could be duplicates). A higher ratio will mean fewer, more efficient cleanings but will mean more wasted space in the log.</td>
-    </tr>
-    <tr>
-      <td>min.insync.replicas</td>
-      <td>1</td>
-      <td>min.insync.replicas</td>
-      <td>When a producer sets acks to "all", min.insync.replicas specifies the minimum number of replicas that must acknowledge a write for the write to be considered successful. If this minimum cannot be met, then the producer will raise an exception (either NotEnoughReplicas or NotEnoughReplicasAfterAppend).
```
Comment (jjkoshy) [inline docs/configuration.html:109]: Actually I just realized that these "hand-written" configurations are more descriptive so we should probably additionally update the `LogConfig` class with these.

Comment (granthenke) [inline docs/configuration.html:109]: @jjkoshy Are you planning to update the descriptions in this PR?

Comment (jjkoshy) [inline docs/configuration.html:109]: Yes - will do that sometime this week.

Comment (jjkoshy) [inline docs/configuration.html:109]: Obviously has conflicts due to Becket's patch, but I'll leave this open for review given that the scope became a little larger (i.e., to consolidate broker and topic configurations). So not sure if everyone will be on board with this approach. cc a few others as well @junrao @becketqin to solicit thoughts/concerns.

**Comment by gwenshap** [comment]
LGTM

**Comment by gwenshap** [comment]
Ah, sorry - looks like there are merge conflicts. @jjkoshy, mind rebasing?

**Comment by jjkoshy** [comment]
@gwenshap @becketqin one observation is that most of the log-specific configs in `KafkaConfig` are _namespaced_ with `log.`. Although we did not do that for `unclean.leader.election.enable`, `compression.type` and `min.insync.replicas` I think we can deal with those later. However for the newer ones from KIP-3x (`message.*`) configs we should probably namespace those in `KafkaConfig` under `log.` before the next release.

**Comment by gwenshap** [comment]
Completely agree, @jjkoshy. Having a list of name conversions from cluster config to logs is embarrassing :)

**Comment by jjkoshy** [comment]
Actually that turns out to be more involved than it seems. I was hoping to just prefix _all_ the log configs in `KafkaConfig` with log and the `LogConfig` doc would simply say that the server defaults are the same but with a prefix of log. Unfortunately some of the base names are different so I would have to _deprecate_ all of those. E.g., `flush.ms` in `LogConfig` `flush.interval.ms` in `KafkaConfig`; and even trickier: `log.roll.jitter.{ms,hours}` in `KafkaConfig` vs `segment.jitter.ms`

These changes just ensure that the server log configs are prefixed with log., but do not eliminate the need for a conversion table.

**Comment by ijuma** [comment]
@jjkoshy I filed https://issues.apache.org/jira/browse/KAFKA-3373 for the KIP-31/32 configs. We should make a decision on that before 0.10.0.0 even if this particular PR is not ready. However, we could fix that as part of this PR if it is ready.

**Comment by becketqin** [comment]
@jjkoshy Good point. I will add the 'log' prefix.

**Comment by jjkoshy** [comment]
So here is a more elaborate/invasive patch - it basically deprecates all the non-conforming configs.

**Comment by jjkoshy** [comment]
I would have liked this to go into 0.10.0 but I'm not sure it is feasible given that the latest patch needs a review and deprecates a number of configurations. I haven't done this in the current patch but if everyone agrees with the general approach then I would also emit a formal warning about the deprecated configurations if specified in the broker configuration or topic-level configuration (via the `TopicCommand` utilities).

**Comment by ijuma** [comment]
@jjkoshy I agree that it's a bit tricky to get this in given the number of deprecations. I think it probably needs to be discussed in the mailing list. Would a KIP be appropriate?

**Comment by jjkoshy** [comment]
Yes it should probably involve a (a rather boring) KIP

**Comment by ijuma** [comment]
:)

**Comment by vvcephei** [comment]
Hi @jjkoshy ,

It seems like this PR stalled. I'll close it out for now, but if you or anyone else want to resume this work, please feel free to re-open it (or start a new one)!

Thanks,
John

**Pull request closed without merging**